asllib/Typing.ml

```diff
@@ -735,6 +735,269 @@ module Annotate (C : ANNOTATE_CONFIG) : S = struct
+
+    (** Returns the slice of the sequence of indices given by [indices],
+        defined by the relative positions given by [slices].
+    *)
+    let slice_indices indices slices =
+      list_concat_map (slice_indices_by_slice [] indices) slices
+      |> coalesce_intervals
+
+    (** Returns the list of absolute bitfields for the bitfield [bf]
+        given that [absolute_parent] is the absolute bitfield
+        for the bitfield where [bf] is declared.
+    *)
+    let rec bitfield_to_absolute env bf absolute_parent =
+      let { abs_name; abs_slices } = absolute_parent in
+      let bf_name = List.append abs_name [ bitfield_get_name bf ] in
+      let slices_as_indices =
+        List.map (slice_to_indices env) (bitfield_get_slices bf)
+      in
+      let bf_indices = slice_indices abs_slices slices_as_indices in
+      let bf_absolute = { abs_name = bf_name; abs_slices = bf_indices } in
+      let nested = bitfield_get_nested bf in
+      bf_absolute :: bitfields_to_absolute env nested bf_absolute
+
+    (** Returns the list of absolute bitfields for the bitfields [bitfields]
+        given that [absolute_parent] is the absolute bitfield
+        where [bitfields] are declared.
+    *)
+    and bitfields_to_absolute env bitfields absolute_parent =
+      list_concat_map
+        (fun bf -> bitfield_to_absolute env bf absolute_parent)
+        bitfields
+
+    (** Tests whether absolute fields [f1] and [f2] are aligned.
+        If the two fields don't share a name or don't exist in the same
+        scope, the result is true.
+    *)
+    let absolute_fields_align f1 f2 =
+      if same_name f1 f2 && exist_in_same_scope f1 f2 then
+        let { abs_slices = indices1 } = f1 in
+        let { abs_slices = indices2 } = f2 in
+        list_equal ( = ) indices1 indices2
+      else true
+
+    (* Begin TypingRule.CheckCommonBitfieldsAlign *)
+    let check annot env bitfields width =
+      if width = 0 then
+        (* We assume that the containing bitvector was successfully annotated.
+           Therefore, if the bitvector type is the empty bitvector,
+           the annotation already ensures its bitfields are aligned. *)
+        ()
+      else
+        let () =
+          if false then
+            Format.eprintf "CheckCommonBitfieldsAlign.check %a@ %a@ %i@."
+              StaticEnv.pp_env env PP.pp_bitfields bitfields width
+        in
+        let () = assert (width > 0) in
+        (* define a fake absolute field representing the entire bitvector. *)
+        let top_absolute = { abs_name = []; abs_slices = [ (0, width - 1) ] } in
+        let absolute_bitfields =
+          bitfields_to_absolute env bitfields top_absolute
+        in
+        let () =
+          if false then
+            List.iter
+              (fun f ->
+                Format.eprintf "absolute field %a@." pp_absolute_bitfield f)
+              absolute_bitfields
+        in
+        iter_ordered_pairs
```

Comment:
I was intentionally avoiding any algorithmic improvements, since the expected size of the problem is a small constant. My goal was to make the implementation as close as possible to the declared requirement.